design-proposals/vm-memory-snapshot.md

```diff
@@ -0,0 +1,151 @@
+
+## Repos
+[KubeVirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+As mentioned the memory snapshot can be taken either in the Online VMSnapshot or as a stand alone command.
+To trigger the memory snapshot during the Online VMSnapshot one will add a parameter to the VMSnapshot yaml.
+To trigger a memory snapshot as a stand alone command a virtctl memoryDump command will be added.
+
+** Before doing either option a pvc should be mounted to the virt-launcher pod. A volume source type will be added to indicate that this is a volume without a matching disk since the VMI doesn't need to access the memory dump, only the virt launcher. This volume source will be able to be added to the VM spec as a volume or by using the Hot-plug API that will be extended to adding this volume without a disk.
+Only one memoryDump volume will be for the VM, the content will be overwritten in every memory dump call.
+
+## API
+
+### New volume source
+
+This volume source will wrap a PVC and will be called memoryDumpVolumeSource. It should be clear that this volume source will *not* have a matching disk and should be handled appropriately. It will be mounted to a hard coded mounting point in the virt launcher pod.
+The definition look like this:
+
+```golang
+// MemoryDumpVolumeSource represents a volume without a disk that will store a future memory dump
+// this volume is only mounted to the virt launcher pod
+type MemoryDumpVolumeSource struct {
+    PersistentVolumeClaim *v1.PersistentVolumeClaimVolumeSource `json:"persistentVolumeClaim,omitempty"`
+}
+```
+
+### Add the new volume source to a VM
+
+#### Add to VM yaml
+
+One way to add this volume source should be by adding it to the vm yaml for example:
+```yaml
+---
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  name: my-vm
+spec:
+  running: true
+  template:
+    spec:
+      domain:
+        devices:
+          disks:
+          - disk:
+              bus: virtio
+            name: containerdisk
+          - disk:
+              bus: virtio
+            name: cloudinitdisk
+        resources:
+          requests:
+            memory: 512Mi
+      terminationGracePeriodSeconds: 0
+      volumes:
+      - containerDisk:
+          image: registry:5000/kubevirt/fedora-with-test-tooling-container-disk:devel
+        name: containerdisk
+      - memoryDump:
+          persistentVolumeClaim:
+              claimName: memory-dump
+        name: disk1
+      - cloudInitNoCloud:
+          userData: |-
+            #cloud-config
+            password: fedora
+            chpasswd: { expire: False }
+        name: cloudinitdisk
+```
+
+This can be added before creating the VM, or if the VM already exists need to add and restart the VM to make the pvc mounted.
+
+#### Add memory dump volume as a hotplug
+
+Use the hot-plug API to add a memory dump volume by adding a flag to the `virtctl addVolume` command:
+
+`$ virtctl addvolume my-vm --volume-name=memory-dump --memoryDumpVolume`
```

Comment:
I don't understand why this has to be tied to disks or volumes.

Reply:
The idea is that there is a "memory" volume type but the volume is not to be included in the list of disks since it should not be visible to the vm.  But could understand it not being a "volume" and just some other part of VM/VMI configuration.  If it is just a volume it may fit in easier with VirtualMachineExport proposal which exports volumes as http endpoints

But I think the PVC target for memory dumps should be configured somewhere on the VM/VMI.  Especially when you consider the hibernate use case.  Do you agree with that?

Reply:
As I mentioned in the previous comment the dump needs to be a part of the VM cause it should be available also after VM shutdown so it cant only be a part of the VMI status for example.
It was considered to the the create of the pvc and binding as part of the dump command but since it should also happen in the snapshot it seems like it will add too much latency to the snapshot process.

Reply:
You can also have stuff in the VM status. But I really don't see how adding "special" volumes is in any way convenient for users and consistent with the spec, and I also don't see the need for it without further explanations.


I think you can try to tackle this from a few different angles.

Firs, if a memory dump is really done on the VM/VMI it is very likely not a declarative action. Adding special volumes does not make it more declarative. So you anyway already have an imperative trigger to do the actual dump.

I then see multiple ways, depending on how much work you want to happen on the VM or the VMI. Like, if you do an imperative /dump call on the VM, you could in that call either specify a PVC, or let the VM controller just create a PVC and record the procgress in the status.

The VM controller could then trigger on the VMI level a /dump call where it insructs the VMI to dump into that PVC. Everything can be visibly tracked on the VM and VMI status and the latest snapshot can be found any time from the VM status.

This is of course just one possibility. What do you think?

Reply:
We looked at doing something similar to disk hotplug and the trigger for virt-handler to add a volume to the virt-launcher pod is a new volume being added to the VMI object. This is where the idea of a non disk volume comes from. We basically need a spot to deposit the memory dump and a volume seems like a natural spot. Now the mechanism which we use to add the volume is up for debate, we just started with something similar to hotplug.

Reply:
If I am not mistaken, when we do hot plug we add a volume because we want a volume, and when we want it gone we remove the volume. This is declarative and we are not adding a pseudo volume from an architectural perspective. Or maybe I misunderstood what you meant?

I think I laid out a few other possibilities, did you also look into something like them?

Reply:
I suggested initially to do the entire process as part of a single command. Mount the volume, do the memory dump, unmount the volume. Which is a prescriptive operation. But if we think of the add on functionality of taking a memory dump together with a VM snapshot, or a hibernate type scenario, those are both declarative operations, and having a prescriptive operation be part of those declarative operations might not be the best idea. Which is why are trying to see if it makes sense (or is possible) for this to be a declarative operation.

Reply:
VM snapshot can definitely be done in a declarative state from a VMSnapshot object point of view. I would not do a live snapshot on a VM in a declarative way. I don't even think that is possible.

For hibernate, yes hibernate can be expressed declarative on the VM, and it can use in the background common flows with a "debug" memory dump.